### PR TITLE
Observable types

### PR DESCRIPTION
**Pull request by ShaMan123**
## Motivation

Enforce strict typing for events handlers in the Observables.

## Description

~~This PR showcases the pros of TS.~~
~~With it using the observable is so easy and fun.~~
~~Found a few discrepancies in the code and fixed them.~~

This PR add a type definition file that describes in minute detail every event fired by fabricJS.
Events are divided by kind, transformation events, selection events, canvas/group events ( add and remove from the object tree), and interaction events.

Every class then takes a dynamic type of event type that is used to bring down to Observable class the type of event that will fire.

Doing so when using the on function for example, Typescript will be able to suggest what events are available and what options are available in the callback.

There are no functional changes on top of that, apart some event property change that has been found clashing with types and so fixed.
The changed events are:
- after:render, added the ctx property to the callback
- selection:cleared, fixed the last instance that was using target instead of deselected
- before:selection:cleared, changed from target to deselected ( those were all `target` but it seems to me that `deselected` makes more sense )

## Changes

- fix some events that were fired with missing/wrong data 

## In Action

ezgif com-gif-maker (4)

### CODE REVIEW

File: src/EventTypeDefs.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    // selection
+    selected: never;
+    deselected: never;
+
+    // tree
+    added: { target: Group | Canvas };
+    removed: { target: Group | Canvas };
+
+    // IText
+    // todo - move to file
```
Comment (ShaMan123) [inline src/EventTypeDefs.ts]: depends on #8420

File: src/mixins/object_interactivity.mixin.ts [hunk 2]
```diff
@@ -19,7 +20,9 @@ type TOCoord = IPoint & {
 
 type TControlSet = Record<string, Control>;
 
-export class InteractiveFabricObject extends FabricObject {
+export class InteractiveFabricObject<
+  T extends TObjectEvents = TObjectEvents
+> extends FabricObject<T> {
```
Comment (ShaMan123) [inline src/mixins/object_interactivity.mixin.ts:25]: made it generic for future use, when #8420 is merged I will move the text events to there so TS won't show they are possible to listen to on other objects (same for group)

File: src/shapes/group.class.ts [hunk 3]
```diff
@@ -3,6 +3,36 @@ import { Point } from '../point.class';
...
+  | 'layout_change'
+  | 'imperative';
+
+export type LayoutContext = {
+  type: LayoutContextType;
+  /**
+   * array of objects starting from the object that triggered the call to the current one
+   */
+  path?: Group[];
+  [key: string]: any;
```
Comment (ShaMan123) [inline src/shapes/group.class.ts:21]: todo in follow up

File: src/typedefs.ts [hunk 4]
```diff
@@ -131,7 +131,7 @@ export type TEvent<E extends Event = TPointerEvent> = {
   e: E;
 };
 
-export type TransformEvent = TEvent & {
+export type TransformEvent<E extends Event = TPointerEvent> = TEvent<E> & {
```
Comment (ShaMan123) [inline src/typedefs.ts:134]: I think all the types related to events should be extracted from this file
Didn't want to cause too many merge conflicts in parallel PRs

File: src/shapes/itext.class.ts [hunk 5]
```diff
@@ -1,6 +1,13 @@
-//@ts-nocheck
+// @ts-nocheck
+import { TObjectEvents, XTransformEvent } from '../EventTypeDefs';
 import { FabricObject } from './fabricObject.class';
 
+export type ITextEvents = TObjectEvents & {
```
Comment (ShaMan123) [inline src/shapes/itext.class.ts:5]: todo in #8420

**Review by ShaMan123 — COMMENTED** [review]
`src/canvas.class.ts` contains the fixes done in the PR

**Review by ShaMan123 — COMMENTED** [review]
extracted all event types from `src/typedefs.ts` to `src/EventTypeDefs.ts`

File: src/EventTypeDefs.ts [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+export type TEvent<E extends Event = TPointerEvent> = {
+  e: E;
+};
+
+export type BasicTransformEvent<E extends Event = TPointerEvent> = TEvent<E> & {
+  transform: Transform;
+  pointer: Point;
+};
+
+export type TModificationEvents =
```
Comment (ShaMan123) [inline src/EventTypeDefs.ts:69]: Up to here are types from typedefs

**Review by ShaMan123 — COMMENTED** [review]
READY to MERGE

**Comment by asturur** [comment]
@ShaMan123 i changed the description of the PR with the one i would have loved to find when i approached the PR.
That would have save me the first 15 minutes of `shit shit shit shit what is all this shit`.
Scrolling lines of combined types doesn't magically tell me your intentions on how you wanted to type those things and why, on a pr of 26 files my process is:

- first check everything quickly
- then close away the files that have no meaningful changes
- take notes on the other so i avoid to leave comments that maybe are clarified by some other file later
- then when i have a grasp of what is going on reset my expectations and actually comment on what is really just important/meaningful

this process requires time, an introduct really helps

File: src/canvas.class.ts [hunk 7]
```diff
@@ -441,10 +441,12 @@ import { saveObjectTransform } from './util/misc/objectTransforms';
         this._objectsToRender = undefined;
         // removing active object should fire "selection:cleared" events
         if (obj === this._activeObject) {
-          this.fire('before:selection:cleared', { target: obj });
+          this.fire('before:selection:cleared', { deselected: [obj] });
```
Comment (asturur) [inline src/canvas.class.ts:444]: So here i m thorn between deselected and maybeDeselected or something like that.
Typescript allow for self descovery of data structure, that means that we can care a bit less about consistency ( all events have target! ) and have names that make sense.
In this case deselected is past, while the before:selection:cleared can be interrupted and is not deselected yet.

Ideas?

File: src/EventTypeDefs.ts [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        };
+
+    // IText
+    'text:selection:changed': { target: FabricObject };
+    'text:changed': { target: FabricObject };
+
+    // misc
+    'contextmenu:before': SimpleEventHandler<Event>;
+    contextmenu: SimpleEventHandler<Event>;
+  };
```
Comment (asturur) [inline src/EventTypeDefs.ts:222]: this is useful, yes, but needs to be a suggestion and not a lockdown.
The developer needs to be able to do

`rect.fire('whateveriwant', ({ ...myStuff }) => { })`

I didn't try the branch, is this possible or locked?

**Comment by asturur** [comment]
I left the only important comment i have, i would like to merge this work, meanwhile i ll search some other PR to merge, if i don't get answers ( and that is totally fine ) if tomorrow i have time i ll try to figure out the answer and in case fix it.

File: src/canvas.class.ts (hunk 7 above)
Comment (ShaMan123) [reply src/canvas.class.ts:444]: I get your point. A good point indeed

Comment (ShaMan123) [reply src/canvas.class.ts:444]: MaybeDeselcted is bad naming. Target or desecltionTarget are better

File: src/EventTypeDefs.ts (hunk 8 above)
Comment (ShaMan123) [reply src/EventTypeDefs.ts:222]: TS will complain.
I can look into it

Comment (ShaMan123) [reply src/EventTypeDefs.ts:222]: I think that if we accept string in addition to the event names TS won't show autocomplete.

**Comment by ShaMan123** [comment]
Great description, point taken

**Comment by ShaMan123** [comment]
I am taking a look now

**Comment by ShaMan123** [comment]
Seems I managed to enable passing other events that are not part of the spec, with an additional overload to `on|once` signature.
However it seems that using the methods with an obj doesn't autocomplete the keys, but does autocomplete the callback. 

```ts
import { FabricObject } from "./fabricObject.class";
import { Group } from "./group.class";

new FabricObject({}).on('resizing',opt=>opt.)
new FabricObject({}).on('',opt=>opt.)
new FabricObject({}).on('resizing1',(opt)=>opt.)
new Group({}).once('layout',opt=>opt.)
new Group({}).on('dgdfg', (opt: { a: boolean }) =>opt.)
new Group({}).on({
    'mousedown:before':opt=>opt.transform,
    'mousedown':opt=>opt.button,
    'added': opt => opt.target,
    
    
})
```

**Comment by ShaMan123** [comment]
Added IText event spec.
Good enough to merge.

**Comment by asturur** [comment]
Gave a stab at solving conflicts

**Comment by ShaMan123** [comment]
Joining now.
Group needs some care as well

**Comment by ShaMan123** [comment]
This PR is an example why #8394 is important
Now after fixing conflicts I would have wanted a test to tell that types work (autocomplete) as expected

**Comment by asturur** [comment]
> This PR is an example why #8394 is important Now after fixing conflicts I would have wanted a test to tell that types work (autocomplete) as expected

I understand that.
But i m thorn between having something that is another thing to test rather than a support.
We are going to add type tests, i m really looking at things to complete the migration before adding more stuff on top.

To be honest i don't even expect this kind of complexity in types, i don't think TS should be a replacemente for documentation.

**Comment by ShaMan123** [comment]
> We are going to add type tests, i m really looking at things to complete the migration before adding more stuff on top.

This is vague

**Comment by asturur** [comment]
> > We are going to add type tests, i m really looking at things to complete the migration before adding more stuff on top.
> 
> This is vague

Meaning that i m looking at things in order of priority to close the TS migration.

**Review by asturur — DISMISSED** [review]

File: src/EventTypeDefs.ts [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+export type ObjectPointerEvents = TPointerEvents<'mouse'>;
+export type CanvasPointerEvents = TPointerEvents<'mouse:'>;
+
+export type ObjectEvents = ObjectPointerEvents &
+  DnDEvents &
+  ObjectModifiedEvents & {
+    // selection
+    selected: never;
+    deselected: never;
```
Comment (asturur) [inline src/EventTypeDefs.ts:175]: to me looks like that selected and deselected do get `target` and optionally 'e' ( if they happen by mouse )

Comment (ShaMan123) [reply src/EventTypeDefs.ts:175]: I recall that the object is fired without context
But need to double check

**Comment by ShaMan123** [comment]
I think we should merge and come back to it if necessary

File: src/EventTypeDefs.ts (hunk 9 above)
Comment (ShaMan123) [reply src/EventTypeDefs.ts:175]: As opposed to the events fired on canvas

Comment (asturur) [reply src/EventTypeDefs.ts]: I verified and added e + target.

**Review by asturur — APPROVED** [review]

**Pull request merged**